drivers/led/lp50xx.c

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ * Copyright (c) 2022 Grinn
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief LP50xx LED controller
+ */
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/drivers/led/lp50xx.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(lp50xx);
+
+#include "led_context.h"
+
+/* Common registers */
+#define LP50XX_DEVICE_CONFIG0		0x00
+#define LP50XX_DEVICE_CONFIG1		0x01
+#define LP50XX_LED_CONFIG0		0x02
+
+/* LP5009/12 registers */
+#define LP5012_LED0_BRIGHTNESS		0x07
+#define LP5012_OUT0_COLOR		0x0b
+
+/* LP5018/24 registers */
+#define LP5024_LED0_BRIGHTNESS		0x07
+#define LP5024_OUT0_COLOR		0x0f
+
+/* LP5030/36 registers */
+#define LP5036_LED_CONFIG1		0x03
```

Comment:
This register is used for bank management of LEDs. It is the same as `LP50xx_LED_CONFIG0`. I know bank management is not available for now, but should I leave these two definitions outside ?

Reply:
Since bank management is only available for apps using the channel API then maybe we should move them into `include/zephyr/drivers/led/lp50xx.h`.

Reply:
I think we are not talking about the same thing. When I use the term _bank management_, I am referring to a specific feature of the chipset where multiple LEDs could be driven as one. This would require an update to the devicetree to handle a child node with a `reg` array (like Linux does in the corresponding driver).

This will also have an impact on the channel API because it will change the base address to use and/or it will introduce some gaps when using the channel API (mapping is bank_brightness, bank_red, bank_green, bank_blue, led[0;n]_brightness, led0_red, led0_green, led0_blue, ..., ledn_red, ledn_green, ledn_blue). That is way I kept the feature disabled for now.

Reply:
We are talking about the same thing :) I am well aware of the bank management feature of the LP503{0,6} controllers. There is actually no support for this feature in the LED API. So the bank management registers are useless in the driver itself.

Before this PR the bank registers were exposed via the `write_channels` API. That's why I said that their definitions belong to `include/zephyr/drivers/led/lp50xx.h`. But I now realize it is no longer the case and it is quite a change... Please explain a bit more why we can't keep that.

Reply:
> We are talking about the same thing :) I am well aware of the bank management feature of the LP503{0,6} controllers.
OK, I just wanted to avoid misunderstandings.
> There is actually no support for this feature in the LED API. So the bank registers definitions are useless in the driver itself.
Using the bank registers should be possible with a `reg` array in the devicetree presented as a single LED through the LED API. This is what the Linux driver is doing and I think that should work in Zephyr too. But I think this feature should be added in another PR since the extension of this driver has waited long enough.
> Before this PR the bank registers were exposed via the `write_channels` API. That's why I said that their definitions belong to `include/zephyr/drivers/led/lp50xx.h`. But I now realize it is no longer the case and it is quite a change... Please explain a bit more why we can't keep that.
I did not understand that the `write_channels()` API allowed to use the bank registers. But since it is a raw access, I get your point. However, I think the problem with this API is that it relies too much on the knowledge of the chip and that will make an application using this driver not really portable. But I may have understood wrongly.

If it is too much a change, I can try to find a way to make it work again. Or find another way to make the bank feature available.